api/public/public_api/endpoints/wake_word_file.py

```diff
@@ -0,0 +1,169 @@
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+"""Public Device API endpoint for uploading a sample wake word for tagging."""
+from datetime import datetime
+from http import HTTPStatus
+from logging import getLogger
+from os import environ
+from pathlib import Path
+
+from flask import jsonify
+from schematics import Model
+from schematics.types import StringType
+from schematics.exceptions import DataError
+
+from selene.api import PublicEndpoint
+from selene.data.account import Account, AccountRepository
+from selene.data.tagging import (
+    TaggingFileLocationRepository,
+    WakeWordFile,
+    WakeWordFileRepository,
+)
+from selene.data.wake_word import WakeWordRepository
+
+LOCAL_IP = "127.0.0.1"
+
+_log = getLogger(__package__)
+
+
+class UploadRequest(Model):
+    """Data class for validating the content of the POST request."""
+
+    wake_word = StringType(required=True)
+    engine = StringType(required=True)
+    timestamp = StringType(required=True)
+    model = StringType(required=True)
+
+
+class WakeWordFileUpload(PublicEndpoint):
+    """Endpoint for submitting and retrieving wake word sample files.
+
+    Samples will be saved to a temporary location on the API host until a daily batch
+    job moves them to a permanent one.  Each file will be logged on the sample table
+    for their location and classification data.
+    """
+
+    _file_location = None
+    _wake_word_repository = None
+    _wake_word = None
+
+    def __init__(self):
+        super(WakeWordFileUpload, self).__init__()
+        self.request_data = None
+
+    @property
+    def wake_word_repository(self):
+        """Lazy instantiation of wake word repository object."""
+        if self._wake_word_repository is None:
+            self._wake_word_repository = WakeWordRepository(self.db)
+
+        return self._wake_word_repository
+
+    @property
+    def wake_word(self):
+        """Build and return a WakeWord object."""
+        if self._wake_word is None:
+            self._wake_word = self.wake_word_repository.ensure_wake_word_exists(
+                name=self.request_data["wake_word"].strip(),
+                engine=self.request_data["engine"],
+            )
+
+        return self._wake_word
+
+    @property
+    def file_location(self):
+        """Build and return a TaggingFileLocation object."""
+        if self._file_location is None:
+            data_dir = Path(environ["SELENE_DATA_DIR"])
+            wake_word = self.request_data["wake_word"].replace(" ", "-")
+            wake_word_dir = data_dir.joinpath("wake-word").joinpath(wake_word)
+            wake_word_dir.mkdir(parents=True, exist_ok=True)
+            file_location_repository = TaggingFileLocationRepository(self.db)
+            self._file_location = file_location_repository.ensure_location_exists(
```

Comment:
What happens if the location does not exist?

Reply:
The mkdir() funtion will create the directory on the server and the ensure_location_exists function will create the directory on the DB.

Reply:
But what happens if mkdir fails?